### PR TITLE
Fix Mysql protocol when using parallel formats (CSV / TSV)

### PR DESCRIPTION
**Pull request by Algunenano**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix Mysql protocol when using parallel formats (CSV / TSV)

Detailed description / Documentation draft:

While looking at why 01176_mysql_client_interactive was broken I found 3 things:
* The "mysql>" prefix is different for mariadb (mariadb 10.6.3), where it's something like `MySQL [system]>` so I've adapted it to make it work.
* The amount of read rows isn't returned.
* The query would get stuck when doing `select * from system.one format TSV;` instead of throwing an error. This is because `TSV` would use a parallel output and there CH doesn't check that it's not using Mysql (as it does in the general path).

I've tested this with mariadb (10.6.3) and mysql (8.0.25) clients and it worked fine with the new regex. I have no idea if this goes through CI and what mysql version it's being used there, but the fact that it was broken tells me it's probably isn't part of the CI.

### CODE REVIEW

File: src/Formats/FormatFactory.cpp [hunk 1]
```diff
@@ -314,6 +314,10 @@ OutputFormatPtr FormatFactory::getOutputFormatParallelIfPossible(
 
     auto format_settings = _format_settings ? *_format_settings : getFormatSettings(context);
 
+    /// If we're handling MySQL protocol connection right now then MySQLWire is only allowed output format.
+    if (format_settings.mysql_wire.sequence_id && (name != "MySQLWire"))
```
Comment (alexey-milovidov) [inline src/Formats/FormatFactory.cpp:318]: Ok but... maybe move this check to somewhere near MySQLHandler?
We can check for existence of FORMAT clause in AST and throw exception if any format is specified.

Comment (Algunenano) [reply src/Formats/FormatFactory.cpp:318]: Makes sense. I'll look for a good place to unify both checks.

**Comment by nikitamikhaylov** [comment]
@Algunenano does the issue reproduce if you set output_format_parallel_formatting= 0 ?

**Comment by Algunenano** [comment]
> @Algunenano does the issue reproduce if you set output_format_parallel_formatting= 0 ?

No, in that case everything works fine and you get the expected error:

```
MySQL [(none)]> select * from system.one format TSV settings output_format_parallel_formatting=0;
ERROR 1 (00000): Code: 1. DB::Exception: MySQL protocol does not support custom output formats. (UNSUPPORTED_METHOD) (version 21.9.1.1)
MySQL [(none)]> 
```

**Comment by Algunenano** [comment]
I couldn't find a better place to handle this than FormatFactory.cpp. The MySQLHandler doesn't know anything about the query AST as it passes the string to executeQuery and that has 2 main paths, both calculating the output format in the same way, that end in getOutputStreamParallelIfPossible or getOutputFormatParallelIfPossible.

What I can do is to keep the old check and disable parallel output formats when the MYSQL sequence_id is declared. That way there is only one check for format compability.

Another option would be to parse the query in MySQLHandler.cpp but that would require either copying a lot of the logic from executeQuery or parsing the query twice, both of which seem like a worse option. An even worse option would be, in my opinion, to try to detect a FORMAT clause in the query string (like the replacements).

I'm going to use the first option. Let me know if you prefer any of the other options or see a better alternative.

**Comment by Algunenano** [comment]
The stress test is complaining about columnNullable:

```
2021.08.09 17:38:29.822230 [ 245802 ] {} <Fatal> BaseDaemon: ########################################
2021.08.09 17:38:29.822443 [ 245802 ] {} <Fatal> BaseDaemon: (version 21.9.1.7712, build id: CE1E0E82A02345E5854F3C20F96C5F4EA3478645) (from thread 7257) (query_id: 86afe420-5875-4e64-8685-96d7c5dffaf3) Received signal Segmentation fault (11)
2021.08.09 17:38:29.822605 [ 245802 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
2021.08.09 17:38:29.822731 [ 245802 ] {} <Fatal> BaseDaemon: Stack trace: 0x22e79c07 0x21352f95 0x22695dcd 0x2428e05d 0x242898ac 0x2427912e 0x2429e6dd 0x2b5335af 0x2b534173 0x2b80dc15 0x2b807f37 0x7f9d0e353609 0x7f9d0e27a293
2021.08.09 17:38:29.886566 [ 245802 ] {} <Fatal> BaseDaemon: 3. ./obj-x86_64-linux-gnu/../src/Columns/ColumnNullable.cpp:640: DB::ColumnNullable::checkConsistency() const @ 0x22e79c07 in /usr/bin/clickhouse
2021.08.09 17:38:29.936017 [ 245802 ] {} <Fatal> BaseDaemon: 4. ./obj-x86_64-linux-gnu/../src/DataTypes/Serializations/SerializationNullable.cpp:74: DB::SerializationNullable::serializeBinaryBulkWithMultipleStreams(DB::IColumn const&, unsigned long, unsigned long, DB::ISerialization::SerializeBinaryBulkSettings&, std::__1::shared_ptr<DB::ISerialization::SerializeBinaryBulkState>&) const @ 0x21352f95 in /usr/bin/clickhouse
2021.08.09 17:38:29.969397 [ 245802 ] {} <Fatal> BaseDaemon: 5.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:2844: std::__1::shared_ptr<DB::ISerialization const>::operator->() const
2021.08.09 17:38:29.980452 [ 245802 ] {} <Fatal> BaseDaemon: 5.2. inlined from ../src/DataStreams/NativeBlockOutputStream.cpp:61: DB::writeData(DB::IDataType const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn> const&, DB::WriteBuffer&, unsigned long, unsigned long)
2021.08.09 17:38:29.980597 [ 245802 ] {} <Fatal> BaseDaemon: 5. ../src/DataStreams/NativeBlockOutputStream.cpp:126: DB::NativeBlockOutputStream::write(DB::Block const&) @ 0x22695dcd in /usr/bin/clickhouse
2021.08.09 17:38:30.334025 [ 245802 ] {} <Fatal> BaseDaemon: 6. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:0: DB::TCPHandler::sendData(DB::Block const&) @ 0x2428e05d in /usr/bin/clickhouse
2021.08.09 17:38:30.714063 [ 245802 ] {} <Fatal> BaseDaemon: 7. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:0: DB::TCPHandler::processOrdinaryQueryWithProcessors() @ 0x242898ac in /usr/bin/clickhouse
2021.08.09 17:38:31.062810 [ 245802 ] {} <Fatal> BaseDaemon: 8. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:354: DB::TCPHandler::runImpl() @ 0x2427912e in /usr/bin/clickhouse
2021.08.09 17:38:31.591939 [ 245802 ] {} <Fatal> BaseDaemon: 9. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:1672: DB::TCPHandler::run() @ 0x2429e6dd in /usr/bin/clickhouse
2021.08.09 17:38:31.597893 [ 245802 ] {} <Fatal> BaseDaemon: 10. ./obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x2b5335af in /usr/bin/clickhouse
2021.08.09 17:38:31.629577 [ 245802 ] {} <Fatal> BaseDaemon: 11. ./obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerDispatcher.cpp:114: Poco::Net::TCPServerDispatcher::run() @ 0x2b534173 in /usr/bin/clickhouse
2021.08.09 17:38:32.128676 [ 245802 ] {} <Fatal> BaseDaemon: 12. ./obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:0: Poco::PooledThread::run() @ 0x2b80dc15 in /usr/bin/clickhouse
2021.08.09 17:38:32.156989 [ 245802 ] {} <Fatal> BaseDaemon: 13. ./obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread_POSIX.cpp:0: Poco::ThreadImpl::runnableEntry(void*) @ 0x2b807f37 in /usr/bin/clickhouse
2021.08.09 17:38:32.172169 [ 245802 ] {} <Fatal> BaseDaemon: 14. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
2021.08.09 17:38:32.172304 [ 245802 ] {} <Fatal> BaseDaemon: 15. __clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.08.09 17:38:32.854921 [ 245802 ] {} <Fatal> BaseDaemon: Calculated checksum of the binary: 53606593028A5647C33C854094313BD4. There is no information about the reference checksum.
2021.08.09 17:38:52.771687 [ 435 ] {} <Fatal> Application: Child process was terminated by signal 11.
```

This seems to be happening in master CI too (e.g. https://clickhouse-test-reports.s3.yandex.net/0/9cbc4b4f7fd7481202dee93fce59e320abb4e2e4/stress_test_(address)/test_run.txt.out.log

OTOH, the Asan fuzzer run is failing due to an OOM in **the client**:
```sql
SELECT
    topKResampleState(1048576, 100, 1048576, 1025)(toString(number), number),
    NULL,
    toTypeName(topKState(1024)(toString(number)))
FROM numbers(-0.)
```

This query is fast but seems to require lots of memory in the client (100GB+ at least), but I don't know why. In any case, not related to this change either.

**Comment by alexey-milovidov** [comment]
Yes, we have to put everything on pause until this is fixed in master...

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by alexey-milovidov** [comment]
@Mergifyio update

**Comment by alexey-milovidov** [comment]
Ok.

@Algunenano I'm reading this PR again without any prior context and I wondering:
- why CSV/TSV even work for MySQL protocol (as I thought it should only use MySQLWire format);
- why disabling parallel formatting for them if MySQL protocol is used is a proper change?

**Comment by alexey-milovidov** [comment]
@Algunenano 

We can replace "ship in the bottle" version of `executeQuery` in MySQLHandler.cpp to the version with AST and BlockIO,
and then check for non-default format (or simply replace all formats to MySQLWire in AST level).

**Comment by Algunenano** [comment]
> We can replace "ship in a bottle" version of executeQuery in MySQLHandler.cpp to the version with AST and BlockIO,
and then check for non-default format (or simply replace all formats to MySQLWire in AST level).

I've just noticed that executeQuery (one of them at least, and it's the one being used MySQLHandler) has a set_result_details callback that passes the format as an argument. Using that I think we could move the check directly to `MySQLHandler.cpp` which as you said is a much better place.

**Comment by Algunenano** [comment]
Changed in https://github.com/ClickHouse/ClickHouse/pull/27609

**Pull request merged**